### ISSUE TITLE
When cross compiling Rust to Android on Windows, Rust 1.69.0+ can't find the version script path.

### ISSUE THREAD

**Issue by moe-deer** [rust-lang/rust#113711]
On Windows, when cross compiling Rust code to generate an Android .so library, using Rust versions 1.69.0 or newer can result in compile errors about the version script.
The error happens because the Rust compiler tries to pass a path to a version script to the linker, but that specified path does not actually exist on Windows.

I tried this code:

```rust
cargo build --release --target=aarch64-linux-android
```

I expected to see this happen: 
```
Finished release [optimized] target(s) in 2.44s
```

Instead, this happened: 
```
error: linking with `D:/NDK/android-ndk-r25c/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android30-clang.cmd` failed: exit code: 255                                                                
  |
  = note: "D:/NDK/android-ndk-r25c/toolchains/llvm/prebuilt/windows-x86_64/bin/aarch64-linux-android30-clang.cmd" "-Wl,--version-script=C:\\Users\\Liste\\AppData\\Local\\Temp\\rustcPwRX2e\\list" "-Wl,--no-undefi
ned-version" "C:\\Users\\Liste\\AppData\\Local\\Temp\\rustcPwRX2e\\symbols.o" "E:\\idea_projects\\my_lib\\target\\aarch64-linux-android\\release\\deps\\my_lib.my_lib.d56b3372c49eacad-cgu.0.rcgu.o" "
-Wl,--as-needed" "-L" "E:\\idea_projects\\my_lib\\target\\aarch64-linux-android\\release\\deps" "-L" "E:\\idea_projects\\my_lib\\target\\release\\deps" "-L" "C:\\Users\\Liste\\.rustup\\toolchains\\stable-
x86_64-pc-windows-msvc\\lib\\rustlib\\aarch64-linux-android\\lib" "-Wl,-Bstatic" "C:\\Users\\Liste\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\aarch64-linux-android\\lib\\libcompiler_builtins-0
1f0855226f8e564.rlib" "-Wl,-Bdynamic" "-ldl" "-llog" "-lunwind" "-ldl" "-lm" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "C:\\Users\\Liste\\.rustup\\toolchains\\stable-x86_64-pc-windows-msvc\\lib\\rustlib\\a
arch64-linux-android\\lib" "-o" "E:\\idea_projects\\my_lib\\target\\aarch64-linux-android\\release\\deps\\libmy_lib.so" "-Wl,--gc-sections" "-shared" "-Wl,-z,relro,-z,now" "-Wl,--strip-all" "-nodefaultlibs"
  = note: --version-script=C:\Users\Liste\AppData\Local\Temp\rustcPwRX2e\list"" was unexpected at this time.


warning: `my_lib` (lib) generated 2 warnings                                                                                                                                                                     
error: could not compile `my_lib` (lib) due to previous error; 2 warnings emitted
```
### Meta

`rustc --version --verbose`:
```
rustc 1.71.0 (8ede3aae2 2023-07-12)
binary: rustc
commit-hash: 8ede3aae28fe6e4d52b38157d7bfe0d3bceef225
commit-date: 2023-07-12
host: x86_64-pc-windows-msvc
release: 1.71.0
LLVM version: 16.0.5
```

**Comment by jyn514** [comment]
you are using an MSVC host toolchain and the compiler is passing unix-style flags. do you see this same problem on windows-gnu or only on MSVC?

**Comment by moe-deer** [comment]
Yes, it only happens on Windows. Linux and Mac work correctly. I'm sure I configured the NDK tools correctly, because it works properly in rust 1.68.2 version.

**Comment by moe-deer** [comment]
I encountered this issue on both windows-gnu and MSVC. They work correctly in rust 1.68.2 version, but the problem appears in versions after that.

**Comment by apiraino** [comment]
@moe-deer Does it also happen with a new Rust project? Any chance you can provide a small bare Rust project template to reproduce the issue? Just to be sure I understand: is the correct Windows path to your Android NDK `D:\\NDK\\android-ndk-r25c\\`?

**Comment by jing332** [comment]
This is because the cmd script cannot pass special symbols such as  `,`  `=` as parameters

**Comment by jing332** [comment]
Error: `version-script"" was unexpected at this time`
![image](https://github.com/rust-lang/rust/assets/42014615/65a60f87-cb87-4b3e-bf1d-cf2518409934)

---
When I remove the `space`, `=`, `,` it can be passed normally:


![image](https://github.com/rust-lang/rust/assets/42014615/36942c37-082f-4a0f-991c-c8f32c7500ec)

**Comment by knopp** [comment]
> Version:
> 
> ```
> rustc 1.71.0 (8ede3aae2 2023-07-12)
> binary: rustc
> commit-hash: 8ede3aae28fe6e4d52b38157d7bfe0d3bceef225
> commit-date: 2023-07-12
> host: x86_64-pc-windows-msvc
> release: 1.71.0
> LLVM version: 16.0.5
> ```
> 
> This is because the cmd script cannot parse the parameter
> 
> Error: `version-script=xx"" was unexpected at this time` ![image](https://user-images.githubusercontent.com/42014615/257474585-65a60f87-cb87-4b3e-bf1d-cf2518409934.png)
> 
> When I remove the `space`, `=`, `,` it can be passed normally:
> 
> ![image](https://user-images.githubusercontent.com/42014615/257474918-36942c37-082f-4a0f-991c-c8f32c7500ec.png)

Neither `"W, version-script=xx"` nor `"-Wversion-script"` in your case would be proper linker flags. Linker flags are prefixed with `-Wl,`.

The problem is this line in the clang script:
```
if "%1" == "-cc1" goto :L
```

It escapes the first argument, but the argument is already escaped which causes problem in the condition and for some reason cmd unescapes and splits the argument and then complains that the second argument is unexpected.

**Comment by knopp** [comment]
In other words, the invocation only works as long as first the linker argument doesn't have comma in it.

**Comment by knopp** [comment]
This seems similar enough, just a different separator: https://github.com/android/ndk/issues/1856.

**Comment by soupslurpr** [comment]
I am getting this exact error, and this workaround works https://github.com/android/ndk/issues/1856#issuecomment-1542248775

**Comment by usernamety** [comment]
Has the problem been solved now?

**Comment by knopp** [comment]
> Has the problem been solved now?

This is not a Rust error. It is a problem with the CMD wrapper in Android SDK. I believe cargo-ndk is not calling the wrapper script anymore.

**Comment by darcycui** [comment]
> This seems similar enough, just a different separator: [android/ndk#1856](https://github.com/android/ndk/issues/1856).

This solved my problem,thanks very much!
